java/src/main/java/com/google/oak/client/OakClient.java

```diff
@@ -0,0 +1,129 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package com.google.oak.client;
+
+import java.util.Optional;
+
+/**
+ * Abstract client for sending and receiving encrypted messages to a server. This class provides an
+ * abstract builder that must be implemented by concrete subclasses of this class. The builder takes
+ * a {@code RpcClientProvider}, an {@code EncryptorProvider}, and optionally an {@code
+ * EvidenceProvider} to build a concrete instance of {@code OakClient}. If an evidence supplier is
+ * provided, the client has to verify the correctness of the evidence prior to getting an instance
+ * of {@code Encryptor} and sending messages to the server. In a complete, and production-ready
+ * implementation, an {@code EvidenceProvider} MUST be provided and verified.
+ *
+ * @param R type of the requests that this client sends
+ * @param T type of the responses that this client receives
+ */
+public abstract class OakClient<R, T> implements AutoCloseable {
+  // TODO(#3063): Replace return type with Result<T>
+  /**
+   * Sends a request to a remote server and receives the response.
+   * @param request the request to send to the server
+   * @return the response received from the server wrapped in an Optional
+   */
+  abstract Optional<T> send(final R request);
```

Comment:
> What is the difference between this method and the one on the `RpcClient` interface? AFAICT this class could be an instance of `RpcClient` itself, with the difference that this class would always deal with plaintext data, while the inner `RpcClient` instance (from the provider) would normally handle encrypted data.

Subclasses of `OakClient` are not expected to directly implement `RpcClient`. This method acts as a wrapper or adapter and delegates the actual RPC call to an instance of `RpcClient`. I think changing the types of request and response to byte array in `RpcClient` can better reflect this. This class is responsible for encryption and decryption of messages and any other transformation.  
 
> BTW, it would also be nice to align with the terminology we are using about IDL / invocations. For instance, this may be called `Transport` or `Channel` and expose a single `invoke` method, and at that point it also becomes the same abstraction that we have elsewhere (which also means we could use the IDL to auto generate code around it in the future, if we want to -- but even if not, it seems worth thinking about aligning the terminology and mental model).

I don't think that is a good idea. The IDL sounds like an implementation detail to me, is at a lower level of abstraction than what I am aiming for in `OakClient` and `RpcClient`, and seems to be evolving at a different pace than this transport-agnostic API should. We will have an implementation of `RpcClient` that uses IDL and that should be aligned with the terminology as well.